client/consensus/grandpa/src/environment.rs

```diff
@@ -1307,14 +1307,46 @@ where
 		.or_else(|| Some((target_header.hash(), *target_header.number()))))
 }
 
+/// Whether we should process a justification for the given block.
+///
+/// This can be used to decide whether to import a justification (when
+/// importing a block), or whether to generate a justification from a
+/// commit (when validating). Justifications for blocks that change the
+/// authority set will always be processed, otherwise we'll only process
+/// justifications if the last one was `justification_period` blocks ago.
+pub(crate) fn should_process_justification<BE, Block, Client>(
+	client: &Client,
+	justification_period: u32,
+	number: NumberFor<Block>,
+	enacts_change: bool,
+) -> bool
+where
+	Block: BlockT,
+	BE: BackendT<Block>,
+	Client: ClientForGrandpa<Block, BE>,
+{
+	if enacts_change {
+		return true
+	}
+
+	let last_finalized_number = client.info().finalized_number;
+
+	// keep the first justification before reaching the justification period
+	if last_finalized_number.is_zero() {
+		return true
+	}
+
+	last_finalized_number / justification_period.into() != number / justification_period.into()
```

Comment:
No, this is not equivalent because we are doing integer division, e.g. given `justification_period = 10`, `last_finalized_number = 1` and `number = 2` this should return `false`.